## Implementing a routine to compute vector length

### Preliminaries

<p><font color=red> Again, copy this notebook so that you don't corrupt the original!  Then you can "play" with the copy of the notebook all you want! </font> </p>

<p>
<b>
NOTE: A common problem that students have with Jupyter notebooks is not understanding that when the code in the gray boxes (cells) is executed, it assigns variables that persist the whole time that the notebook is open. Further, some cells rely on variables assigned by earlier cells.  If you execute these cells out of order, or if you execute the same cell twice, then you may end up changing the value of the variables.  To correct this, click on "Cell" at the top and execute "run all above" or "run all".  You can also reset all cells by clicking "Cell -> All Output -> Clear"
</b>
</p>

<p>In this notebook, we show how to write a routine to compute vector length. </p>

Let's start by creating a vector $ x = \left( \begin{array}{r} 1 \\ 2 \\ 3 \end{array} \right) $.

Execute the code in the box by clicking in the box and then on "Cell -> Run".  Alternative, click on the box and push "Shift" and "Return" (or "Enter") together.

In [1]:
# create a column vector x. 
x = [1, 2, 3]
@show x;

x = [1, 2, 3]


Now, recall that the length of a vector equals $ \| x \|_2 = \sqrt{ x^T x } $.  So, we can use the dot product routine that we wrote before to compute the length:

In [34]:
include("../laff/laff.jl")
using .laff

In [35]:
length_x = sqrt( laff.dot( x, x ) )

@show length_x;

length_x = 3.7416573867739413


### Length as a simple routine

<p>
As before, it is preferable to create a routine that computes the length.
</p>

<p>
Complete the following routine to implement this:
</p>

In [41]:
function length( x )
    return sqrt( laff.dot( x, x ) )
end

length (generic function with 1 method)

Be sure the run the above box, or this notebook won't know about the routine!!!

Now, if you execute

In [45]:
length_x = length( x )

println( "length_x:" )
println( length_x )

println( "difference between length_x and sqrt( laff.dot( x, x ) ):" )
println( length_x - sqrt( laff.dot( x, x ) ) )

length_x:
3.7416573867739413
difference between length_x and sqrt( laff.dot( x, x ) ):
0.0


The result should be:

<code>
length_x:
3.7416573867739413
difference between length_x and sqrt( laff.dot( x, x ) ):
0.0
</code>

### A complete length function as part of the LAFF library

Note that your implementation of `length` will already work on column and row (transposed column) vectors.

In [46]:
import Base.print_array

In [47]:
x = transpose([1, 2, 3])

println("x is a row vector")
println("x = ")
print_array(stdout, x)

length_x = length(x)
println("\nlength_x = ")
println(length_x)

x is a row vector
x = 
 1  2  3
length_x = 
3.7416573867739413


A routine similar to what you've written for this exercise is part of the 'laff' library.  If you do

```julia
include("../laff/laff.jl")
using .laff
```
then <code> laff.norm2( x ) </code> will perform the desired computation of the length of $ x $, when <code> x </code> is a column and/or a row vectors.  If you really want to see what this routine looks like, then ask for it on the discussion forum and we'll point you to where it can be found.

In [57]:
include("../laff/laff.jl")
using .laff

In [58]:
length_x = laff.norm2( x )

println( "length_x:" )
println( length_x )

println( "difference between length_x and sqrt( laff.dot( x, x ) ):" )
println( length_x - sqrt( laff.dot( x, x ) ) )

length_x:
3.741657386773941
difference between length_x and sqrt( laff.dot( x, x ) ):
-4.440892098500626e-16


### Need a challenge?

In "1.7.3 Overflow and Underflow", we discuss how computing the length with the dot product can inadvertently cause overflow or underflow.  Write a routine that avoids that

In [ ]:
function length( x )
    ### You fill in the rest!
end